### PR TITLE
[TD-5062] Feature/row partial update

### PR DESCRIPTION
**Pull request by cosmic-jc**

### CODE REVIEW

File: src/tsdb/inc/tsdbRowMergeBuf.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "tsdb.h"
+
+typedef void* SMergeBuf;
+
+#define SMERGE_BUF_LEN(x) (*(int*)(x))
```
Comment (hzcheng) [inline src/tsdb/inc/tsdbRowMergeBuf.h]: In osMemory.h, we have a taosTMalloc series interfaces which can provide same functions as these. Besides, we have tsdbMakeRoom() function. Try to use those interfaces.

File: src/tsdb/src/tsdbRowMergeBuf.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "tsdbRowMergeBuf.h"
+#include "tdataformat.h"
+
+int tsdbMergeBufMakeSureRoom(SMergeBuf *pBuf, STSchema* pSchema1, STSchema* pSchema2) {
+  int len = MAX(dataRowMaxBytesFromSchema(pSchema1), dataRowMaxBytesFromSchema(pSchema2)) + sizeof(int);
```
Comment (hzcheng) [inline src/tsdb/src/tsdbRowMergeBuf.c]: If you really want to use a memory allocation strategy as this, make this function FORCE_INLINE maybe a good choice.

File: src/tsdb/src/tsdbRowMergeBuf.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    return NULL;
+  }
+  SMergeBuf buf = *pBuf;
+
+  void *pData = SMERGE_BUF_PTR(buf);
+
+  return mergeTwoMemRows(pData, row1, row2, pSchema1, pSchema2);
+
+  /*
+  *(uint16_t*)pData = *(uint16_t*)row1;
```
Comment (hzcheng) [inline src/tsdb/src/tsdbRowMergeBuf.c]: Delete these comment lines. If you want to keep some code as your logical flow tips, use #if 0 -- #endif.

File: src/tsdb/src/tsdbMain.c [hunk 4]
```diff
@@ -106,6 +107,17 @@ STsdbRepo *tsdbOpenRepo(STsdbCfg *pCfg, STsdbAppH *pAppH) {
     return NULL;
   }
 
+  if(pCfg->update == TD_ROW_PARTIAL_UPDATE) {
+    SMergeBuf buf;
+    if((buf = tsdbMakeBuf(512)) == NULL) {
+      tsdbError("vgId:%d failed to open TSDB repository while opening merge buffer pool since %s", config.tsdbId,
```
Comment (hzcheng) [inline src/tsdb/src/tsdbMain.c]: We can avoid allocating memory here by using taosTMalloc series APIs. Allocate memory when it is really neccessary.

File: src/tsdb/src/tsdbMemTable.c [hunk 5]
```diff
@@ -693,95 +702,190 @@ static int tsdbScanAndConvertSubmitMsg(STsdbRepo *pRepo, SSubmitMsg *pMsg) {
...
+}
 
-  return 0;
+static void* tsdbInsertDupKeyMergePacked(void** args) {
+  return tsdbInsertDupKeyMerge(args[0], args[1], args[2], (STSchema**)&args[3], (STSchema**)&args[4], args[5]);
+}
 
-_err:
-  return -1;
+static void tsdbSetupSkipListHookFns(SSkipList* pSkipList, STsdbRepo *pRepo, STable *pTable, int32_t* affectedRows, int64_t* points) {
```
Comment (lichuang) [inline src/tsdb/src/tsdbMemTable.c]: 看了一下一类函数的定义，看起来每一类函数的参数数量都是固定的，既然如此，就不用在创建的时候传入参数数量了，且这类函数的arg也不要写数组数量为0的变参了。

File: src/util/inc/tfunctional.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+//TODO: hard to use, trying to rewrite it using va_list
+
+typedef void* (*GenericVaFunc)(void* args[]);
+typedef int32_t (*I32VaFunc) (void* args[]);
+typedef void (*VoidVaFunc) (void* args[]);
+
+typedef struct GenericSavedFunc {
+  GenericVaFunc func;
+  void * args[];
```
Comment (lichuang) [inline src/util/inc/tfunctional.h:32]: 每一类函数的参数数量是固定的，这里可以不用使用变参来保存参数了吧，下面几个函数定义也是一样的问题。

File: src/util/src/tskiplist.c [hunk 7]
```diff
@@ -31,11 +32,33 @@ static SSkipListNode *tSkipListNewNode(uint8_t level);
 static SSkipListNode *tSkipListPutImpl(SSkipList *pSkipList, void *pData, SSkipListNode **direction, bool isForward,
                                        bool hasDup);
 
+
 static FORCE_INLINE int     tSkipListWLock(SSkipList *pSkipList);
 static FORCE_INLINE int     tSkipListRLock(SSkipList *pSkipList);
 static FORCE_INLINE int     tSkipListUnlock(SSkipList *pSkipList);
 static FORCE_INLINE int32_t getSkipListRandLevel(SSkipList *pSkipList);
 
+void FORCE_INLINE tSkipListPutBatchSetHookFns(SSkipList* pSkipList,
```
Comment (hzcheng) [inline src/util/src/tskiplist.c]: If this function is not used by other files, make it static is better.

File: src/util/inc/tfunctional.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+} tVoidSavedFunc;
+
+tGenericSavedFunc* genericSavedFuncInit(GenericVaFunc func, int numOfArgs);
+tI32SavedFunc* i32SavedFuncInit(I32VaFunc func, int numOfArgs);
+tVoidSavedFunc* voidSavedFuncInit(VoidVaFunc func, int numOfArgs);
+void* genericInvoke(tGenericSavedFunc* const pSavedFunc);
+int32_t i32Invoke(tI32SavedFunc* const pSavedFunc);
+void voidInvoke(tVoidSavedFunc* const pSavedFunc);
+
+#define FREE_SAVED_FUNC(x)                \
```
Comment (hzcheng) [inline src/util/inc/tfunctional.h]: We have a tfree() macro which performs the same role as this macro. Use that one to get rid of duplicate codes.

File: src/util/inc/tskiplist.h [hunk 9]
```diff
@@ -95,6 +106,19 @@ typedef struct tSkipListState {
   uint64_t nTotalElapsedTimeForInsert;
 } tSkipListState;
 
+typedef enum {
+  SSkipListPutBatchPreNormal    = 0,
+  SSkipListPutBatchPreEarlyStop = 1
+} SSkipListPutBatchPreStatus;
+
+#define INCLUDE_PUTBATCH_HOOK()  \
+  tI32SavedFunc* preBatchFn;     \
```
Comment (hzcheng) [inline src/util/inc/tskiplist.h]: Try to simplify the hook functions.

File: src/util/inc/tfunctional.h (hunk 6 above)
Comment (lichuang) [reply src/util/inc/tfunctional.h:32]: 另外，最好在代码里注释一下每个参数都保存的是什么值

File: src/util/src/tskiplist.c [hunk 10]
```diff
@@ -124,6 +155,92 @@ SSkipListNode *tSkipListPut(SSkipList *pSkipList, void *pData) {
...
+  bool           hasDup = false;
+  char *         pKey = NULL;
+  char *         pDataKey = NULL;
+  int            compare = 0;
+
+  tSkipListWLock(pSkipList);
+
+  void* pData = iterate(iter);
+  if(pData == NULL) return;
+  pSkipList->preEntryFn->args[2] = pData;
```
Comment (lichuang) [inline src/util/src/tskiplist.c]: 下面178行有判断pSkipList->preEntryFn为空的逻辑，所以pSkipList->preEntryFn不是必然不为空的吧，这里是不是也要判断一下才能赋值

File: src/util/src/tskiplist.c [hunk 11]
```diff
@@ -124,6 +155,92 @@ SSkipListNode *tSkipListPut(SSkipList *pSkipList, void *pData) {
...
+    voidInvoke(pSkipList->postEntryFn);
+  }
+
+  for (int level = 0; level < pSkipList->maxLevel; level++) {
+    forward[level] = SL_NODE_GET_BACKWARD_POINTER(backward[level], level);
+  }
+
+  // forward to put the rest of data
+  while ((pData = iterate(iter)) != NULL) {
+    pSkipList->preEntryFn->args[2] = pData;
```
Comment (lichuang) [inline src/util/src/tskiplist.c]: 下面178行有判断pSkipList->preEntryFn为空的逻辑，所以pSkipList->preEntryFn不是必然不为空的吧，这里是不是也要判断一下才能赋值

File: src/util/src/tskiplist.c [hunk 12]
```diff
@@ -124,6 +129,79 @@ SSkipListNode *tSkipListPut(SSkipList *pSkipList, void *pData) {
...
 }
 
+void tSkipListPutBatchByIter(SSkipList *pSkipList, void *iter, iter_next_fn_t iterate, void **lastData) {
+  SSkipListNode *backward[MAX_SKIP_LIST_LEVEL] = {0};
+  SSkipListNode *forward[MAX_SKIP_LIST_LEVEL] = {0};
+  bool           hasDup = false;
+  char *         pKey = NULL;
+  char *         pDataKey = NULL;
+  int            compare = 0;
+  SSkipListNode* pNode;
```
Comment (hzcheng) [inline src/util/src/tskiplist.c]: No other place using tSkipListPutBatch, just change that API to make minimum changes.

File: src/tsdb/src/tsdbMemTable.c [hunk 13]
```diff
@@ -693,95 +701,183 @@ static int tsdbScanAndConvertSubmitMsg(STsdbRepo *pRepo, SSubmitMsg *pMsg) {
...
-  void *         rows[TSDB_MAX_INSERT_BATCH] = {0};
-  int            rowCounter = 0;
-
-  ASSERT(pBlock->tid < pMeta->maxTables);
-  pTable = pMeta->tables[pBlock->tid];
-  ASSERT(pTable != NULL && TABLE_UID(pTable) == pBlock->uid);
+#if 0
+static FORCE_INLINE int32_t insertDataToTablePreEntry(STsdbRepo* pRepo, STable *pTable, SMemRow row) {
+  STsdbCfg *  pCfg = &pRepo->config;
+  TKEY        tkey = memRowTKey(row);
```
Comment (hzcheng) [inline src/tsdb/src/tsdbMemTable.c]: Are these code still need to be kept?

File: src/util/inc/tfunctional.h (hunk 6 above)
Comment (cosmic-jc) [reply src/util/inc/tfunctional.h:32]: 这里本来用partial applied function这种东西，但C里面没有，且比较难做到。因为我需要保存参数和函数指针，并调用它，如果不使用void * args[]，那么就需要把所有参数都写在结构体里。为了避免每定义一个partial applied都需要定义一个结构体的麻烦，就抽象出了void * func(void*[] args)这样一个函数，然后把相关参数扔到结构体里面去。这样只需要在使用的时候指定参数个数并设置参数，调用的时候也可以用一个通用的invoke函数，用起来容易一些。
每个参数的值是有说明的，但不再这里，因为这是一个抽象的东西。通常在使用它的地方都会有说明。

**Review by lichuang — APPROVED** [review]

File: src/tsdb/src/tsdbRead.c [hunk 14]
```diff
@@ -3834,8 +3942,8 @@ void tsdbCleanupQueryHandle(TsdbQueryHandleT queryHandle) {
   pQueryHandle->next = doFreeColumnInfoData(pQueryHandle->next);
 
   SIOCostSummary* pCost = &pQueryHandle->cost;
-  tsdbDebug("%p :io-cost summary: head-file read cnt:%"PRIu64", head-file time:%"PRIu64" us, statis-info:%"PRId64" us, datablock:%" PRId64" us, check data:%"PRId64" us, 0x%"PRIx64,
-      pQueryHandle, pCost->headFileLoad, pCost->headFileLoadTime, pCost->statisInfoLoadTime, pCost->blockLoadTime, pCost->checkForNextTime, pQueryHandle->qId);
+  tsdbDebug("%p :io-cost summary: statis-info:%"PRId64" us, datablock:%" PRId64" us, check data:%"PRId64" us, 0x%"PRIx64,
+      pQueryHandle, pCost->statisInfoLoadTime, pCost->blockLoadTime, pCost->checkForNextTime, pQueryHandle->qId);
```
Comment (hjxilinx) [inline src/tsdb/src/tsdbRead.c:3948]: why the statistics information is removed from log？

Comment (cosmic-jc) [reply src/tsdb/src/tsdbRead.c:3948]: I think this is caused by a mistakenly merge conflict handling, before I started this pr, we do not output headFileLoad and headFileLoadTime. Anyway, i'll add it back.

File: src/util/src/tskiplist.c [hunk 15]
```diff
@@ -124,28 +129,33 @@ SSkipListNode *tSkipListPut(SSkipList *pSkipList, void *pData) {
...
+  void* pData = iterate(iter);
+  if(pData == NULL) return;
+
   // backward to put the first data
-  hasDup = tSkipListGetPosToPut(pSkipList, backward, ppData[0]);
-  tSkipListPutImpl(pSkipList, ppData[0], backward, false, hasDup);
+  hasDup = tSkipListGetPosToPut(pSkipList, backward, pData);
+
+  pNode = tSkipListPutImpl(pSkipList, pData, backward, false, hasDup);
+  if(pNode) *lastData = pNode->pData;
```
Comment (hzcheng) [inline src/util/src/tskiplist.c]: The lastData assignment logic should not be in skiplist. It should be in app code. So just put the logic in the callback function.

**Review by hjxilinx — APPROVED** [review]

**Review by hzcheng — APPROVED** [review]

**Pull request merged**